kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -315,8 +500,62 @@ impl<'tcx> GotocCtx<'tcx> {
                     .collect(),
                 &self.symbol_table,
             ),
-            AggregateKind::Adt(_, _, _, _, _) => unimplemented!(),
-            AggregateKind::Closure(_, _) => unimplemented!(),
+            AggregateKind::Adt(_, variant_index, _, _, active_field_index) => {
+                let typ = self.codegen_ty(res_ty);
+                if let Some(field_index) = active_field_index {
+                    // union
+                    assert_eq!(operands.len(), 1);
+                    let components = typ.lookup_components(&self.symbol_table).unwrap();
+                    assert!(components.len() > field_index);
+                    Expr::union_expr(
+                        typ,
+                        components[field_index].name(),
+                        self.codegen_operand(&operands[0]),
+                        &self.symbol_table,
+                    )
+                } else if !res_ty.is_enum() {
+                    // struct
+                    let layout = self.layout_of(res_ty);
+                    Expr::struct_expr_from_values(
+                        typ,
+                        layout
+                            .fields
+                            .index_by_increasing_offset()
+                            .map(|idx| self.codegen_operand(&operands[idx]))
+                            .collect(),
+                        &self.symbol_table,
+                    )
+                } else {
+                    // enum
+                    debug!("IS ENUM");
+                    debug!(?variant_index, "VARIANT INDEX");
+                    let layout = self.layout_of(res_ty);
+                    debug!(?layout, "LAYOUT");
+                    if operands.is_empty() {
+                        self.codegen_rvalue_enum(&variant_index, None, res_ty)
+                    } else {
+                        assert_eq!(operands.len(), 1);
```

Comment:
Why do you have this assertion? I believe the number of operands here will match the number of fields of that variant. For example:
```
enum Enum {
    One(bool),
    Two(bool, bool),
}

fn check() {
    // This will have one operand.
    let _var = Enum::One(false);
    // This will have two operands -- true and false
    let _var = Enum::Two(true, false);
}
```

Reply:
Thank you for this example! I was naively assuming that the `Two` case would result in a single, struct-typed operand.

Reply:
I am now including this as a test case.